include/graaflib/types.h

```diff
@@ -9,14 +9,20 @@ namespace graaf {
 using vertex_id_t = std::size_t;
 using edge_id_t = std::pair<vertex_id_t, vertex_id_t>;
 
+template <class T>
+inline void hash_combine(std::size_t& seed, const T& v) {
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```

Comment:
If I understand correctly, this is [hash_combine of boost v1.33](https://www.boost.org/doc/libs/1_86_0/libs/container_hash/doc/html/hash.html#notes_hash_combine).
A reference to the source as a comment would be good here. The magic number and the shifts otherwise seem a little arbitrary.

Reply:
Ah, it appears that is the source of the hash combine function. I was only familiar with the function based on seeing it used in multiple contexts in production. I hadn't tracked down the source.

It appears that like many magic constants, the value of 0x9e3779b9 was chosen arbitrarily as the fractional component of the golden ratio multiplied by 2^32. Perhaps on modern hardware, using the golden ratio multiplied by 2^64 would show improved performance.

Sources:

http://burtleburtle.net/bob/hash/doobs.html
https://softwareengineering.stackexchange.com/questions/63595/tea-algorithm-constant-0x9e3779b9-said-to-be-derived-from-golden-ratio-but-the/63599#63599

Perhaps it's better in that case to directly use the Boost hash combiner?

Reply:
The bit shifts are probably chosen arbitrarily too.

Reply:
I'll speak with my team at Atomic Industries about making a novel hash combiner function, so we don't accidentally end up plagiarizing another project. We have a few mathematicians on the team with a background in combinatorics.

Reply:
+1 I agree that referencing Boost would be helpful here. Looking at the [Boost Software License](https://www.boost.org/LICENSE_1_0.txt) under which the [ContainerHash](https://github.com/boostorg/container_hash?tab=readme-ov-file) library is licensed this may also be a requirement. No need to push changes to the PR, I can add this before merging :) 

The alternative would be to add Boost as a dependency to Graaf, but I would like to avoid this as I would like Graaf to be a standalone Boost::Graph alternative.